File: Java/242ValidAnagram.java
/*
 * We create a fixed-size array of buckets to hold the occurrences of each character, and check this array to make sure both words have the same number of occurences
 */

public class 242ValidAnagram {
    public boolean isAnagram(String s, String t) {
        char[] b = new char[26];
        for (int i = 0; i < s.length(); i++) {
            b[s.charAt(i) - 'a']++;
        }
        for (int i = 0; i < t.length(); i++) {
            b[t.charAt(i) - 'a']--;
        }
        for (int i = 0; i < 26; i++) {
            if (b[i] != 0) return false;
        }
        return true;
    }
}

File: Python/242ValidAnagram.py
# We create a dict which stores each character and it's number of occurences (+1 for in s, -1 for in t). We then check to make sure the net sum of all characters is 0.

class Solution(object):
    def isAnagram(self, s, t):
        d = {}
        for c in s:
            if c in d:
                d[c] += 1
            else:
                d[c] = 1
        for c in t:
            if c not in d:
                return False
            else:
                d[c] -= 1
        for i in d:
            if d[i] != 0:
                return False
        return True